// interweave/lib/Element.js
"use strict";
/**
 * @copyright   2016-2018, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || Object.assign || function(t) {
    for (var s, i = 1, n = arguments.length; i < n; i++) {
        s = arguments[i];
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
            t[p] = s[p];
    }
    return t;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var react_1 = __importDefault(require("react"));
var prop_types_1 = __importDefault(require("prop-types"));
var Element = /** @class */ (function (_super) {
    __extends(Element, _super);
    function Element() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Element.prototype.render = function () {
        var _a = this.props, attributes = _a.attributes, children = _a.children, selfClose = _a.selfClose, Tag = _a.tagName;
        if (selfClose) {
            return react_1.default.createElement(Tag, __assign({}, attributes));
        }
        return react_1.default.createElement(Tag, __assign({}, attributes), children);
    };
    Element.propTypes = {
        attributes: prop_types_1.default.objectOf(prop_types_1.default.oneOfType([prop_types_1.default.string, prop_types_1.default.number, prop_types_1.default.bool])),
        children: prop_types_1.default.node,
        selfClose: prop_types_1.default.bool,
        tagName: prop_types_1.default.string.isRequired,
    };
    Element.defaultProps = {
        attributes: {},
        children: null,
        selfClose: false,
    };
    return Element;
}(react_1.default.PureComponent));
exports.default = Element;
